### PR TITLE
Add support for account locking

### LINKED ISSUE #14276: Does phpmyadmin support lock/unlock db user feature on web?
### Steps to reproduce
1. I using phpmyadmin 4.8.0.1 latest but I can not find out lock/unlock user on phpmyadmin?
2. Does phpmyadmin support lock/unlock db user feature on web?

### Expected behaviour
There are lock/unlock buttons for every db user (except root) on phpmyadmin
As of version 5.7.6, MySQL supports locking and unlocking user accounts
https://dev.mysql.com/doc/refman/5.7/en/account-locking.html

### PR DESCRIPTION
**Pull request by MauricioFauth**
- MySQL 5.7.6: https://dev.mysql.com/doc/refman/en/account-locking.html
- MariaDB 10.4.2: https://mariadb.com/kb/en/account-locking/

Fixes #14276.

Icons:
- Bootstrap: [lock-open.svg](https://github.com/phpmyadmin/data/blob/master/resources/svg/icons/lock-open.svg)
- Pmahomme and Original: [lock_open.png](http://www.famfamfam.com/lab/icons/silk/)
- Metro: Desaturated version of [lock_open.png](http://www.famfamfam.com/lab/icons/silk/)

### CODE REVIEW

File: libraries/classes/Controllers/Server/Privileges/AccountLockController.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    public function __invoke(ServerRequest $request): void
+    {
+        $userName = (string) $request->getParsedBodyParam('username');
+        $hostName = (string) $request->getParsedBodyParam('hostname');
+        $statement = sprintf(
+            'ALTER USER \'%s\'@\'%s\' ACCOUNT LOCK;',
+            $this->dbi->escapeString($userName),
+            $this->dbi->escapeString($hostName)
+        );
+        $this->dbi->query($statement);
```
Comment (williamdes) [inline libraries/classes/Controllers/Server/Privileges/AccountLockController.php:33]: could we use prepared queries ?

File: libraries/classes/Controllers/Server/Privileges/AccountUnlockController.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace PhpMyAdmin\Controllers\Server\Privileges;
+
+use PhpMyAdmin\Controllers\AbstractController;
+use PhpMyAdmin\DatabaseInterface;
+use PhpMyAdmin\Http\ServerRequest;
+use PhpMyAdmin\ResponseRenderer;
+use PhpMyAdmin\Template;
+
+final class AccountUnlockController extends AbstractController
```
Comment (williamdes) [inline libraries/classes/Controllers/Server/Privileges/AccountUnlockController.php:13]: What about one controller with lock and unlock ?

File: libraries/classes/Server/Privileges.php [hunk 3]
```diff
@@ -2131,6 +2132,8 @@ public function getUsersOverview($result, array $dbRights, $textDir)
         }
 
         $hosts = [];
+        $hasAccountLocking = ($this->dbi->isMariaDB() && $this->dbi->getVersion() >= 100402)
+            || (! $this->dbi->isMariaDB() && $this->dbi->getVersion() >= 50706);
```
Comment (williamdes) [inline libraries/classes/Server/Privileges.php:2136]: This should move to Query\Compatibility

File: themes/bootstrap/img/lock-open.svg [hunk 4]
```diff
@@ -0,0 +1 @@
+<svg viewBox="0 0 32 32" xmlns="http://www.w3.org/2000/svg"><path d="m16 2c-5.5 0-10 4.5-10 10v4h-2v14h24v-14h-2-4-4-8v-4c0-3.3 2.7-6 6-6s6 2.7 6 6h4c0-5.5-4.5-10-10-10zm0 20a2 2 0 0 1 2 2 2 2 0 0 1 -2 2 2 2 0 0 1 -2-2 2 2 0 0 1 2-2z"/></svg>
```
Comment (williamdes) [inline themes/bootstrap/img/lock-open.svg:1]: could you please add some lines to the vendors documentation file with the links you posted about the icons ?

File: js/src/server/privileges.js [hunk 5]
```diff
@@ -325,6 +326,39 @@ AJAX.registerOnload('server/privileges.js', function () {
...
+    $('.js-account-locking').on('click', function () {
+        const button = this;
+        const isLocked = button.dataset.isLocked === 'true';
+        const url = isLocked
+            ? 'index.php?route=/server/privileges/account-unlock'
+            : 'index.php?route=/server/privileges/account-lock';
+
+        $.post(
+            url,
+            { 'username': button.dataset.userName, 'hostname': button.dataset.hostName, 'ajax_request': true },
```
Comment (williamdes) [inline js/src/server/privileges.js:338]: Seems like you are missing the serverId

File: libraries/classes/Controllers/JavaScriptMessagesController.php [hunk 6]
```diff
@@ -137,6 +137,10 @@ private function setMessages(): void
             'strPasswordNotSame' => __('The passwords aren\'t the same!'),
             'strRemovingSelectedUsers' => __('Removing Selected Users'),
             'strClose' => __('Close'),
+            'strLock' => __('Lock'),
+            'strUnlock' => __('Unlock'),
+            'strLockAccount' => __('Lock this account.'),
+            'strUnlockAccount' => __('Unlock this account.'),
```
Comment (williamdes) [inline libraries/classes/Controllers/JavaScriptMessagesController.php:143]: Could you maybe use a context translation or add some l10n details for translators ?

File: test/classes/Server/Privileges/AccountLockingTest.php [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        $dbi->expects($this->once())
+            ->method('getVersion')
+            ->willReturn(100402);
+        $dbi->expects($this->exactly(2))
+            ->method('escapeString')
+            ->withConsecutive([$this->equalTo('test.user')], [$this->equalTo('test.host')])
+            ->willReturnOnConsecutiveCalls('test.user', 'test.host');
+        $dbi->expects($this->once())
+            ->method('tryQuery')
+            ->with($this->equalTo('ALTER USER \'test.user\'@\'test.host\' ACCOUNT LOCK;'))
```
Comment (williamdes) [inline test/classes/Server/Privileges/AccountLockingTest.php:31]: You could have used `$this->dummyDbi->addResult` ;)

**Review by williamdes — APPROVED** [review]
Awesome work !

File: js/src/server/privileges.js (hunk 5 above)
Comment (mauriciofauth) [reply js/src/server/privileges.js:338]: Thanks! We should improve that for JavaScript.

File: test/classes/Server/Privileges/AccountLockingTest.php (hunk 7 above)
Comment (mauriciofauth) [reply test/classes/Server/Privileges/AccountLockingTest.php:31]: `DummyDbi` does not work well with queries that return boolean values.

File: js/src/server/privileges.js (hunk 5 above)
Comment (williamdes) [reply js/src/server/privileges.js:338]: It should be handled like the token, but for some reason it is not
Not sure why, maybe we will find out when fixing multi server login

File: test/classes/Server/Privileges/AccountLockingTest.php (hunk 7 above)
Comment (williamdes) [reply test/classes/Server/Privileges/AccountLockingTest.php:31]: 🤔 I do not see why, maybe I am missing something

**Review by williamdes — APPROVED** [review]
I really like the create stub test way

**Pull request merged**